### PR TITLE
Only register `.jdeps` caching `ActionContext` when used

### PR DESCRIPTION
**Pull request by fmeum**
By registering `JavaCompileActionContext` (renamed to `JdepsCachingActionContext` in this change to better reflect its only purpose) only when using `--java_classpath=bazel`, the `insertDependencies` field of `JavaHeaderCompileAction` becomes obsolete. Combined with the existing 3 bytes of padding, this makes room for a new 4-byte field that will be used in a rollforward of f5d76b1777ce861a7e551342fe25363517d53ff5.

Along the way, skip the unnecessary work of reading the `.jdeps` in `JavaHeaderCompilerAction#afterExecute` if neither using the `.jdeps` cache nor path mapping.

Work towards #21091

### CODE REVIEW

**Comment by fmeum** [comment]
Additional context is provided in https://github.com/bazelbuild/bazel/pull/19872#issuecomment-1937809923.

File: src/main/java/com/google/devtools/build/lib/bazel/rules/BazelRulesModule.java [hunk 1]
```diff
@@ -563,6 +565,10 @@ public void registerActionContexts(
       ModuleActionContextRegistry.Builder registryBuilder,
       CommandEnvironment env,
       BuildRequest buildRequest) {
-    registryBuilder.register(JavaCompileActionContext.class, new JavaCompileActionContext());
+    JavaOptions javaOptions = env.getOptions().getOptions(JavaOptions.class);
+    if (javaOptions != null
```
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/BazelRulesModule.java:569]: The Blaze analogue of this setup would also have to gate adding the context behind this condition.

**Comment by fmeum** [comment]
@bazel-io fork 7.1.0

**Comment by fmeum** [comment]
@hvadehra Btw, does Google use all three modes of `--java_classpath`? I'm wondering whether it would make sense to flip the default for Bazel to `bazel` (from `javabuilder`).

**Comment by hvadehra** [comment]
1. I'm a little unsure about this change. IIUC this change tries to derive the value of `insertDependencies` from the presence on the `JavaCompileActionContext`/`JdepsCachingActionContext`. But those need not be the same, given L374-377. So, after this change, we'd be populating the cache even in those cases where we turn off the `bazel` classpath mode. I don't know if this would matter in practice, but given that this check was explicitly introduced in https://github.com/bazelbuild/bazel/commit/5f9e9f04ecdda8edf779757c7e54aecc29923d91, I'd like @gregestren to have a look as well.

2. I haven't looked at the details of the regression discussed in https://github.com/bazelbuild/bazel/pull/19872#issuecomment-1936866469 but looking at the numbers, I estimate the contribution of the extra field is ~5% of the total regression. This might actually be an acceptable number. I'd recommend tackling the rest first, and then adding this change on if necessary later. In any case, if there is a O(days) gap between merging this change and the rollforward, that latter may still be flagged as a regression since the baseline for measurement will include this change. So, having both ready to be merged together will make things simpler.

3. The default for `--java_classpath` internally is `bazel` as well, and that's the mode used by almost all builds (a small number (0.007%) do use the `javabuilder` mode, but I'll have to investigate to see why). So flipping the default should be mostly a no-op internally and SGTM.

**Comment by hvadehra** [comment]
> Could you link to the lines you referred to as L374-377

I'd meant to do that but forgot, sorry about that: https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/java/JavaHeaderCompileAction.java#L373-L377

**Comment by fmeum** [comment]
Thanks, now I fully understand your comment. This logic does make it difficult to replace the field, assuming it is needed at all.

I will close this PR for now and think about how to prevent the large part of the regression.

I will also look into submitting a PR to flip the default classpath mode for Bazel.

**Pull request closed without merging**